### PR TITLE
DDP-4605 show warnings on mailing address

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
These changes modify API call to verify mailing address.
We can now get "warnings" from /verify
Warnings are now displayed like other errors using ValidationMessage hoping we can make styling and appearance consistent with rest of application.

Updated test to look for warning display.
Separately, discovered some corner cases where similar suggestions and errors were not properly handled.

### CODE REVIEW

File: ddp-workspace/projects/ddp-sdk/src/lib/services/address.service.ts [hunk 1]
```diff
@@ -19,10 +20,10 @@ export class AddressService extends UserServiceAgent<Address> {
...
 
-    public verifyAddress(address: Address): Observable<Address> {
-        return this.postObservable('/profile/address/verify', address, null, true).pipe(
+    public verifyAddress(address: Address): Observable<AddressVerificationResponse> {
+        return this.postObservable('/profile/address/verify', {...address, studyGuid: this.configuration.studyGuid}, null, true).pipe(
             map((data: any) => {
-                return new Address(data.body);
+                return new AddressVerificationResponse(data.body);
             }),
             catchError((error) => {
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/services/address.service.ts:28]: New API call looks for study guid so server can look for study specific rules. In testboston, we look for rules that might trigger zip code warnings. If your zip code not in list, we will warn you.
Response is now a little more complicated. New type needed.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.service.ts [hunk 2]
```diff
@@ -222,38 +216,36 @@ export class AddressInputService implements OnDestroy {
...
-          take(1)
+    const cancelableFormAddress$ = street1Changed$.pipe(withLatestFrom(formAddress$)).pipe(
+      concatMap(([streetChanged, address]) =>
+        of(streetChanged).pipe(
+            delay(streetChanged ? 3000 : 0),
+            mapTo(address),
+            // If autocomplete address has come in during our delay, we don't emit the form address
+            takeUntil(this.googleAutocompleteAddress$),
+            take(1)
         )),
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.service.ts:249]: Noticed that under some circumstances using `zip` was not working as I expected. Driving cancelableFormAddress from street1Changed is more direct and reliable.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 3]
```diff
@@ -287,6 +297,7 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
 
     const verifyInputComponentSparseAddress$ = this.inputComponentAddress$.pipe(
       filter(address => !this.enoughDataToVerify(address)),
+      tap(address => addressSuggestion$.next(null)),
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:304]: Was not hiding suggestions when verifying. Was OK sometimes. Others not.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 4]
```diff
@@ -316,27 +330,44 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+        skip(1),
+        map(formValue => formValue.suggestionRadioGroup as SuggestionOption),
+        distinctUntilChanged(),
+        share());
+
+    const handleSelectedWarnings$ = suggestionRadioValue$.pipe(
+        withLatestFrom(this.suggestionInfo$),
+        tap(([radioValue, suggestionInfo]) =>
+            this.stateUpdates$.next({
+              formWarningMessages: suggestionInfo ? suggestionInfo.warnings[radioValue].map(warn => warn.message) : []
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:369]: The suggestions can sometimes swap in warnings.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 5]
```diff
@@ -257,27 +265,29 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+                  warnings: state.warnings} : null),
       shareReplay(1)
     );
 
+    this.errorMessagesToDisplay$ = this.state$.pipe(
+        map(state => state.formErrorMessages.concat(state.formWarningMessages)),
+        distinctUntilChanged((x, y) => util.isEqual(x, y)),
+        shareReplay(1)
+    );
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:288]: Combine warnings and errors for display. Choosing to display them using the validation message component.

Comment (yufengwng) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:288]: Cool! I think as long as it's not in red, then it should be fine group together the warnings and errors.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 6]
```diff
@@ -186,17 +196,18 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
 
     this.state$ = this.stateUpdates$.pipe(
       startWith(initialState),
       scan((acc: ComponentState, update) => ({ ...acc, ...update })),
+      tap(state =>  console.debug('New embeddedComponentState$=%o', state)),
       shareReplay(1)
     );
-
-    this.state$.subscribe((state) => console.debug('New embeddedComponentState$=' + JSON.stringify(state)));
+    this.state$.subscribe();
```
Comment (fat23cat) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: Hm, do we need to take care of unsubscribing from the observable when we use shareReplay?

Also, could you move this.initializeComponentState(); call in the ngOnInit?

Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts]: I had not unsubscribed properly in this spot and another place. Thnx!
930065a72fac5d10a0edf8522c5c00852929b0ac
Regarding the `initializeComponentState()` as currently configured it has to happen in constructor.

**Review by fat23cat — APPROVED** [review]
Looks good!
